lib/_stream_writable.js

```diff
@@ -496,30 +478,25 @@ function onwrite(stream, er) {
       onwriteError(stream, state, er, cb);
     }
   } else {
-    // Check if we're actually ready to finish, but don't emit yet
-    var finished = needFinish(state) || stream.destroyed;
-
-    if (!finished &&
-        !state.corked &&
-        !state.bufferProcessing &&
-        state.bufferedRequest) {
-      clearBuffer(stream, state);
-    }
-
-    if (sync) {
-      // It is a common case that the callback passed to .write() is always
-      // the same. In that case, we do not schedule a new nextTick(), but rather
-      // just increase a counter, to improve performance and avoid memory
-      // allocations.
-      if (state.afterWriteTickInfo !== null &&
-          state.afterWriteTickInfo.cb === cb) {
-        state.afterWriteTickInfo.count++;
+    clearBuffer(stream, state);
+    if (state.ending || state.needDrain || cb !== nop) {
```

Comment:
Nit: I think `cb` and `needDrain` would be true more often than `ending`, so it'd be better to avoid checking it every time.
```suggestion
    if (cb !== nop || state.needDrain || state.ending) {
```

Reply:
`cb !== nop` is actualyl quite uncommon, but I do agree `needDrain` is more common.